src/game/gui/text_render.c

```diff
@@ -80,27 +89,97 @@ static inline void render_char_surface(const text_settings *settings, text_mode
         default:
             color = settings->cforeground;
     }
+
+    // Draw the actual font
     video_draw_offset(sur, x, y, color - 1, 255);
+    return sur->w;
 }
 
-int text_render_char(const text_settings *settings, text_mode state, int x, int y, char ch) {
-    surface *sur = get_font_surface(settings, ch);
-    render_char_shadow_surface(settings, sur, x, y);
-    render_char_surface(settings, state, sur, x, y);
-    return sur->w;
+void letter_set_parameters(text_object *cached_text, surface *sur, int x, int y, int offset, int limit) {
+    cached_text->cur_letter->x = x;
```

Comment:
Please add an assert that checks that `cached_text->cur_letter < cached_text->letter + cached_text->max_letters`